shelley-ma/formal-spec/introduction.tex

```diff
@@ -1,12 +1,69 @@
-\cite{utxo_ma}. These multi-assets are implemented via token bundles,
-given by a type called $\Value$, which replace the type $\Coin$ of the
-Shelley ledger in many (but not all) cases.
+This document gives a specification of the changes
+required to implement a both an Allegra-era and a Mary-era ledger. These ledgers share
+a specification (and can share an implementation), with the exception of
+a difference in one particular type. We abstract over this type and several
+functions specific to it, which allows us to write a single spec for both versions of the
+ledger, to which we refer as a \emph{ShelleyMA ledger spec}.
 
 All aspects of the ledger not given in this document are as specified
-in the Shelley ledger design and implementation~\cite{shelley_spec},
-but if they depend on something that is changed in this version of the
-ledger they must use the definitions from this document.
+in the Shelley ledger design and implementation~\cite{shelley_spec}.
+Only additions and replacements of definitions of Shelley types, functions, rules, and transitions
+are given here. Replacements are always given the same name as the
+original definition in the Shelley spec, eg. the $\mathsf{UTXO}$ rule in the
+Shelley spec is replaced with the $\mathsf{UTXO}$ rule given here.
+
+\subsection{ShelleyMA, Allegra and Mary Eras}
+
+The additional functionality that is introduced in the Allegra and Mary eras
+is as follows :
+
+\begin{itemize}
+  \item \emph{Allegra} : Timelock scripts replace multi-signature scripts as the native
+  ledger-iterpreted script type. This type of script, in addition to the
+  clauses of the m-of-n multi-signature language, has constructors that
+  allow users to constrain the upper and lower slots of the validity of a
+  clause of the script. For example, one may specify that up to a given slot $s$,
+  either key $k$ or $k'$ can sign the transaction for the script to be valid,
+  whereas from $s$ onwards, only $k$'s signature can make the script valid.
+
+  \item \emph{Mary} : Multi-asset support is added to the Allegra ledger, ie.
+  to the Shelley ledger with timelock scripts. This includes changes to
+  ledger accounting to accommodate multiple types of tokens, as well as
+  support for minting and burning of user-defined tokens. This approach
+  taken here was described initially in \cite{utxo_ma}.
+\end{itemize}
+
+Adding both timelock script and multi-asset support is done in a single specification
+of the ShelleyMA ledger, as the transition rules are exactly the same for both
+Mary and Allegra. The difference between the two is that all accounting is
+done in terms of lovelace, ie. $\Coin$, in the Allegra era, whereas the Mary era
+replaces the $\Coin$ type with a type the terms of which are bundles of arbitrary assets
+(user-defined ones, as well as lovelace), which we call $\Value$.
+
+In this specification, we represent all the places where the $\Coin$ type is used in Allegra,
```

Comment:
I will avoid using Class and Instantiate , but I think using the term LedgerMonoid is more confusing than saying something like "we define an algebraic structure Val, which is a monoid Val with additional functions". In that, I am pretty set on keeping the Val name as it is used in the code (perhaps typeset nicer than it currently is)

Reply:
I don't particularly care about the name, but I think it's nice to have something descriptive. `Val` doesn't really tell you what operations it supports or what it's used for. I wanted to at least have `Monoid` in there so that the reader immediately knows that it's supposed to be a monoid with some extra stuff. Maybe `Val` is also not the best name for the class in the implementation?

Reply:
Coming back to this, `Val` doesn't behave the same grammatically as other algebraic structures. You can say: "Let V be a vector space", but you can't say "Let V be a Val". I'd prefer it a lot if the name would fit in this sentence. Any of the following do: Ledger monoid, Value monoid, Accounting type, Coin-like type, token-bundle type, possibly something else. But having an algebraic structure that doesn't fit in the above sentence is awkward.